geopandas/geodataframe.py

```diff
@@ -1422,6 +1408,35 @@ def apply(self, func, axis=0, raw=False, result_type=None, args=(), **kwargs):
 
     @property
     def _constructor(self):
+        geometry_default = getattr(self, "_geometry_column_name")
+        crs_default = getattr(self, "crs")
+
+        def _geodataframe_constructor_with_fallback(
+            data=None, index=None, crs=crs_default, geometry=geometry_default, **kwargs
+        ):
+            df = pd.DataFrame(data, index=index)
+            if geometry in df.columns:
+                geo_col = df[geometry]
+                if isinstance(geo_col, pd.DataFrame) or isinstance(
+                    geo_col.dtype, GeometryDtype
+                ):
+                    # DataFrame check is for when multiple columns are named
+                    # `geometry`. This case should error, but is passed through so that
+                    # we can have a specific error message for specific contexts,
+                    # e.g. concat, caught in __finalize__.
+
+                    df = GeoDataFrame(
+                        data=data, index=index, crs=crs, geometry=None, **kwargs
+                    )
+                    # Note we cannot supply in constructor or use set_geometry, as
+                    # cached sindex on slices will be reset
+                    df._geometry_column_name = geometry
```

Comment:
I haven't looked at sindex thoroughly, I do wonder if this situation might be handled better by fixing something there, rather than here. This is something I will have another look into.

Reply:
So is this about eg `df["col", "geometry"]]` reusing the same cached sindex if `df` has one cached?

Reply:
Yep, and additionally `self.df.iloc[:]` using cached sindex.  (The tests here check `assert subset1.sindex is original_index` so they're looking for the same object, not the same values)

Reply:
I think this could probably be fixed by set_geometry, doing some kind of early exit if the column already exists of geometrydtype and the crs hasn't been changed.